src/main/kotlin/uk/gov/dluhc/printapi/service/PrintResponseProcessingService.kt

```diff
@@ -0,0 +1,92 @@
+import uk.gov.dluhc.printapi.database.entity.Status.PRINT_PROVIDER_PRODUCTION_FAILED
+import uk.gov.dluhc.printapi.database.entity.Status.PRINT_PROVIDER_VALIDATION_FAILED
+import uk.gov.dluhc.printapi.database.entity.Status.RECEIVED_BY_PRINT_PROVIDER
+import uk.gov.dluhc.printapi.database.entity.Status.SENT_TO_PRINT_PROVIDER
+import uk.gov.dluhc.printapi.database.entity.Status.VALIDATED_BY_PRINT_PROVIDER
+import uk.gov.dluhc.printapi.database.repository.PrintDetailsRepository
+import uk.gov.dluhc.printapi.messaging.models.ProcessPrintResponseMessage
+import uk.gov.dluhc.printapi.printprovider.models.BatchResponse
+import uk.gov.dluhc.printapi.printprovider.models.BatchResponse.Status.SUCCESS
+import java.time.Clock
+import java.time.OffsetDateTime
+import uk.gov.dluhc.printapi.messaging.models.ProcessPrintResponseMessage.Status as ResponseStatus
+import uk.gov.dluhc.printapi.messaging.models.ProcessPrintResponseMessage.StatusStep as ResponseStatusStep
+
+class PrintResponseProcessingService(
+    private val printDetailsRepository: PrintDetailsRepository,
+    private val idFactory: IdFactory,
+    private val clock: Clock
+) {
+    /**
+     * Updates print details for the batches based on the batch responses' statuses.
+     * If a batch is successful, a new printRequestStatus will be added for each corresponding print details with status
+     * being RECEIVED_BY_PRINT_PROVIDER.
+     *
+     * If a batch is not successful, the status will be set to SENT_TO_PRINT_PROVIDER and a new requestId will be set.
+     * The reason for the new requestId is, the print provider does not support duplicate print requests with the same requestId.
+     *
+     * dateCreated will be the timestamp the new printRequestStatus is created and eventDateTime will be the timestamp from
+     * the print provider's batch response.
+     */
+    fun processBatchResponses(batchResponses: List<BatchResponse>) {
+        batchResponses.forEach { batchResponse ->
+            with(batchResponse) {
+                val newStatus =
+                    if (status == SUCCESS) RECEIVED_BY_PRINT_PROVIDER else PENDING_ASSIGNMENT_TO_BATCH
+
+                val printDetails =
+                    printDetailsRepository.getAllByStatusAndBatchId(SENT_TO_PRINT_PROVIDER, batchId)
+
+                printDetails.forEach {
+                    it.addStatus(
+                        status = newStatus,
+                        dateCreated = OffsetDateTime.now(clock),
+                        eventDateTime = timestamp,
+                        message = null
+                    )
+
+                    if (status != SUCCESS) {
+                        it.requestId = idFactory.requestId()
+                    }
+                }
+
+                printDetailsRepository.updateItems(printDetails)
+            }
+        }
+    }
+
+    fun processPrintResponse(printResponse: ProcessPrintResponseMessage) {
+        val newStatus = with(printResponse) {
+            when {
```

Comment:
I don't know if kotlin has pattern matching but can we do  something like 

```
when ((statusStep, status)) {
  (ResponseStatusStep.PROCESSED, ResponseStatus.SUCCESS) -> VALIDATED_BY_PRINT_PROVIDER
}
```

Reply:
Thanks, @fayimora , for the suggestion.
There is actually a way to do it:

<pre>
        val newStatus = with(printResponse) {
            when(statusStep to status) {
                ResponseStatusStep.PROCESSED to ResponseStatus.SUCCESS -> VALIDATED_BY_PRINT_PROVIDER
                ResponseStatusStep.IN_MINUS_PRODUCTION to ResponseStatus.SUCCESS -> IN_PRODUCTION
                ResponseStatusStep.DISPATCHED to ResponseStatus.SUCCESS -> DISPATCHED
                ResponseStatusStep.NOT_MINUS_DELIVERED to ResponseStatus.FAILED -> NOT_DELIVERED
                ResponseStatusStep.PROCESSED to ResponseStatus.FAILED -> PRINT_PROVIDER_VALIDATION_FAILED
                ResponseStatusStep.IN_MINUS_PRODUCTION to ResponseStatus.FAILED -> PRINT_PROVIDER_PRODUCTION_FAILED
                ResponseStatusStep.DISPATCHED to ResponseStatus.FAILED -> PRINT_PROVIDER_DISPATCH_FAILED
                else -> throw IllegalArgumentException("Undefined statusStep [$statusStep] and status [$status] combination")
            }
        }
</pre>

I have read the comments below from @vishalgupta-valtech and @NathanRussellValtech  and seems like it will make it more readable if we deal with success and fail statuses separately, and I agree.

i.e.
<pre>
val newStatus = with(printResponse) {
            when (status) {
                ResponseStatus.SUCCESS -> {
                    when (statusStep) {
                        ResponseStatusStep.PROCESSED -> VALIDATED_BY_PRINT_PROVIDER
                        ResponseStatusStep.IN_MINUS_PRODUCTION -> IN_PRODUCTION
                        ResponseStatusStep.DISPATCHED -> DISPATCHED
                        else -> throw IllegalArgumentException("Undefined statusStep [$statusStep] and status [$status] combination")
                    }
                }
                ResponseStatus.FAILED -> {
                    when (statusStep) {
                        ResponseStatusStep.NOT_MINUS_DELIVERED -> NOT_DELIVERED
                        ResponseStatusStep.PROCESSED -> PRINT_PROVIDER_VALIDATION_FAILED
                        ResponseStatusStep.IN_MINUS_PRODUCTION -> PRINT_PROVIDER_PRODUCTION_FAILED
                        ResponseStatusStep.DISPATCHED -> PRINT_PROVIDER_DISPATCH_FAILED
                    }
                }
            }
        }
</pre>